bundles/org.openhab.binding.novafinedust/src/main/java/org/openhab/binding/novafinedust/internal/SDS011Handler.java

```diff
@@ -0,0 +1,228 @@
+                connectionMonitorStartDelay.getSeconds(), timeBetweenDataShouldArrive.getSeconds(), TimeUnit.SECONDS);
+    }
+
+    private void initializeCommunicator(WorkMode mode, Duration interval) {
+        boolean initSuccessful = false;
+        try {
+            initSuccessful = communicator.initialize(mode, interval);
+        } catch (final IOException ex) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, "I/O error!");
+            return;
+        } catch (PortInUseException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR, "Port is in use!");
+            return;
+        } catch (TooManyListenersException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    "Cannot attach listener to port!");
+            return;
+        } catch (UnsupportedCommOperationException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    "Cannot set serial port parameters");
+            return;
+        }
+
+        if (initSuccessful) {
+            lastCommunication = ZonedDateTime.now();
+            updateStatus(ThingStatus.ONLINE);
+
+            if (mode == WorkMode.POLLING) {
+                pollingJob = scheduler.scheduleWithFixedDelay(() -> {
+                    try {
+                        communicator.requestSensorData();
+                    } catch (IOException e) {
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                                "Cannot query data from device");
+                    }
+                }, 2, config.pollingInterval, TimeUnit.SECONDS);
+            }
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    "Commands and replies from the device don't seem to match");
+            logger.debug("Could not configure sensor -> setting Thing to OFFLINE and disposing the handler");
+            dispose();
+        }
+    }
+
+    private boolean validateConfiguration() {
+        if (config.port == null || config.port.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR, "Port must be set!");
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    public void dispose() {
+        if (pollingJob != null && !pollingJob.isCancelled()) {
+            pollingJob.cancel(true);
+            pollingJob = null;
+        }
+
+        if (connectionMonitor != null && !connectionMonitor.isCancelled()) {
+            connectionMonitor.cancel(true);
+            connectionMonitor = null;
+        }
+
+        if (communicator != null) {
+            scheduler.schedule(() -> communicator.dispose(), 0, TimeUnit.SECONDS);
```

Comment:
Because I think `initialize` and `dispose` from a `ThingHandler` should return quickly, i.e. within 5 seconds. Inside the `dispose` I send the sensor to sleep by communicating with it. This is important to enhance its lifetime.

So in order to not block the disposing call, I do it asynchronously.

Reply:
So what happens if the handler is restarted (disposed then initialized again)? Would the handler be stuck in a configuration COMMUNICATION_ERROR because the port is still in use?

Reply:
This is a VERY interesting point! :-)

There are similar problems in the zigbee binding, because there it also takes a while to startup the stack and if this gets restarted through dispose/initialize calls, everything gets messed up too. Synchronization might help there but is a pain in the ass to do that for every binding.

And I am afraid to say that with the 5 second rule for initialize/dispose, bindings are forced to spawn a separate task, which will ALWAYS cause these troubles if dispose/initialize is called within a short time frame (i.e. by a disable/enable of the thing).

If you prefer to have my dispose here running in sync with the framework, I am fine to replace the task with a direct call. In my case this usually only takes about 1-2 seconds anyway, but I wanted to play safe since I am communicating with an external device and this might timeout, etc.

But the overall problem for all bindings will stay, especially with the use case of enabling and disabling of a thing.

WDYT?

Reply:
The only way I can see of dealing with delayed initialize/dispose is to make your handler and communicator objects have different lifecycles.

You would need to create a your own communicator factory class that handlers would request communicators for a given port. 
```
public interface CommunicatorFactory{
     SDS011Communicator getCommunicator(SerialPortIdentifier portId);
}
```

The factory itself would share the same lifecycle as an osgi component so either you could make your handler factory a communicator factory or implement it in a new osgi component.

The key benefit here is that you would be able to reuse SDS011Communicator instances across handler restarts.

Reply:
This doesn't really solve the problem, I could still have the communicator instance running if i dispose the handler and then the port is blocked.

Also its not worth the effort as I said before the communication takes max 1-2 seconds.

I have changed it to be synchronous now.

Reply:
>This doesn't really solve the problem, I could still have the communicator instance running if i dispose the handler and then the port is blocked.

Not really since you can prevent a communicator from opening the port if it is already open. Likewise a communicator can stop its disposal process if a new handler requests to use it.

>I have changed it to be synchronous now.

I don't think would work, a new handler means a new communicator object. Synchronization only works on the same object instance.

Reply:
> Not really since you can prevent a communicator from opening the port if it is already open. Likewise a communicator can stop its disposal process if a new handler requests to use it.

The problem is not having a different thing, let's say "thing2" wanting the very same port. This is already handled, because the port in use exception is dealt with. And this is always a user problem, since he has configured thing1 and thing2 with the same port.

What I am talking about is that the ThingHandler from thing1 is disposed and shortly afterwards initialized again. This can happen if one disabled and enables a thing, see https://github.com/openhab/openhab-core/blob/5b325aa3d210c2dfa490c676520fe755b0f3fe1c/bundles/org.openhab.core.thing/src/main/java/org/openhab/core/thing/internal/ThingManagerImpl.java#L1167-L1216

Here it doesnt matter if a communicator object or the thinghandler itself want to open the port on initialize. If it is currently in use by a thinghandler that gets disposed at the moment, then there is no way for a starting thinghandler to open the port and he will run into the port in use exception.

This applies to ALL binding which are operating on a shared resource like a port btw...

So with the change I did yesterday evening, the dispose is always called and finished before an initialize, so for now all is good within this binding.

However, one should think about the general problem of shared resources and scenarios where an initialize or dispose might take longer and one has no other choice than putting this in an external task because of this 5 second rule from openHAB.